spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteDelete.scala

```diff
@@ -0,0 +1,152 @@
+
+  import org.apache.spark.sql.execution.datasources.v2.ExtendedDataSourceV2Implicits._
+
+  private val FILE_NAME_COL = "_file"
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
+    // don't rewrite deletes that can be answered by passing filters to deleteWhere in SupportsDelete
+    case d @ DeleteFromTable(r: DataSourceV2Relation, Some(cond)) if isDeleteWhereCase(r, cond) =>
+      d
+
+    // rewrite all operations that require reading the table to delete records
+    case DeleteFromTable(r: DataSourceV2Relation, Some(cond)) =>
+      // TODO: do a switch based on whether we get BatchWrite or DeltaBatchWrite
+      val writeInfo = newWriteInfo(r.schema)
+      val mergeBuilder = r.table.asMergeable.newMergeBuilder(writeInfo)
+
+      val scanPlan = buildScanPlan(r.table, r.output, mergeBuilder, cond)
+
+      val remainingRowFilter = Not(EqualNullSafe(cond, Literal(true, BooleanType)))
+      val remainingRowsPlan = Filter(remainingRowFilter, scanPlan)
+
+      val batchWrite = mergeBuilder.asWriteBuilder.buildForBatch()
+      val writePlan = buildWritePlan(remainingRowsPlan, r.output)
+      ReplaceData(r, batchWrite, writePlan)
+  }
+
+  private def buildScanPlan(
+      table: Table,
+      output: Seq[AttributeReference],
+      mergeBuilder: MergeBuilder,
+      cond: Expression): LogicalPlan = {
+
+    val scanBuilder = mergeBuilder.asScanBuilder
+
+    val predicates = splitConjunctivePredicates(cond)
+    val normalizedPredicates = DataSourceStrategy.normalizeExprs(predicates, output)
+    PushDownUtils.pushFilters(scanBuilder, normalizedPredicates)
+
+    val scan = scanBuilder.build()
+    val scanRelation = DataSourceV2ScanRelation(table, scan, output)
+
+    val scanPlan = scan match {
+      case _: SupportsFileFilter =>
+        val matchingFilePlan = buildFileFilterPlan(cond, scanRelation)
+        val dynamicFileFilter = DynamicFileFilter(scanRelation, matchingFilePlan)
+        dynamicFileFilter
+      case _ =>
+        scanRelation
+    }
+
+    // include file name so that we can group data back
+    val fileNameExpr = Alias(InputFileName(), FILE_NAME_COL)()
+    Project(scanPlan.output :+ fileNameExpr, scanPlan)
+  }
+
+  private def buildWritePlan(
+      remainingRowsPlan: LogicalPlan,
+      output: Seq[AttributeReference]): LogicalPlan = {
+
+    // TODO: sort by _pos to keep the original ordering of rows
+    // TODO: consider setting a file size limit
+
+    val fileNameCol = findOutputAttr(remainingRowsPlan, FILE_NAME_COL)
+    val numShufflePartitions = SQLConf.get.numShufflePartitions
+    val repartition = RepartitionByExpression(Seq(fileNameCol), remainingRowsPlan, numShufflePartitions)
+    val sort = Sort(Seq(SortOrder(fileNameCol, Ascending)), global = false, repartition)
+    Project(output, sort)
+  }
+
+  private def isDeleteWhereCase(relation: DataSourceV2Relation, cond: Expression): Boolean = {
```

Comment:
What about `isMetadataDelete`? I think that's more clear why we would not rewrite the plan. All of these plans are technically `DELETE FROM ... WHERE`.

Reply:
Agree.